ライトセンサトピックの送信周期を上げたい

<!-- 品質向上にご協力ありがとうございます -->
<!-- テンプレートを参考に文章を作成してください -->

# 現状の問題点
<!-- 現状の問題点をここに書きます -->

ライトセンサのトピックは10Hzでパブリッシュされています。

https://github.com/rt-net/raspimouse2/blob/d9df35aa8d9bb2639469150d66e5f8a691ba3927/raspimouse/src/raspimouse_component.cpp#L126-L128

走行しながらセンサ値を処理するために、できるだけセンサ値の送信周期を上げたいです。

# 改善案
<!-- 問題を改善する方法をここに書きます -->

送信周期を10 Hzから100 Hzに上げます。

もしくはROS パラメータで変更できるようにします。

パラメータ化に関してはまだ調査中です。以下現状報告  
  
以下のように100msをget_parameter()関数に置き換えて変更してビルドしてみたところ、
```cpp
  light_sensors_timer_ = create_wall_timer(
    get_parameter(LIGHT_SENSORS_HZ_PARAM).get_value<std::string>(), std::bind(
      &Raspimouse::publish_light_sensors, this));
```
以下のようなエラーが表示されました。
```sh
/home/ubuntu/ros2_ws/src/raspimouse2/raspimouse/src/raspimouse_component.cpp: In member function ‘virtual CallbackReturn raspimouse::Raspimouse::on_configure(const rclcpp_lifecycle::State&)’:
/home/ubuntu/ros2_ws/src/raspimouse2/raspimouse/src/raspimouse_component.cpp:134:48: error: no matching function for call to ‘raspimouse::Raspimouse::create_wall_timer(std::__cxx11::basic_string<char>, std::_Bind_helper<false, void (raspimouse::Raspimouse::*)(), raspimouse::Raspimouse*>::type)’
  134 |       &Raspimouse::publish_light_sensors, this));
      |                                                ^
In file included from /opt/ros/foxy/include/rclcpp_lifecycle/lifecycle_node.hpp:935,
                 from /home/ubuntu/ros2_ws/src/raspimouse2/raspimouse/include/raspimouse/raspimouse_component.hpp:57,
                 from /home/ubuntu/ros2_ws/src/raspimouse2/raspimouse/src/raspimouse_component.cpp:15:
/opt/ros/foxy/include/rclcpp_lifecycle/lifecycle_node_impl.hpp:84:1: note: candidate: ‘template<class DurationRepT, class DurationT, class CallbackT> typename rclcpp::WallTimer<CallbackT>::SharedPtr rclcpp_lifecycle::LifecycleNode::create_wall_timer(std::chrono::duration<_Rep1, _Period1>, CallbackT, rclcpp::CallbackGroup::SharedPtr)’
   84 | LifecycleNode::create_wall_timer(
      | ^~~~~~~~~~~~~
/opt/ros/foxy/include/rclcpp_lifecycle/lifecycle_node_impl.hpp:84:1: note:   template argument deduction/substitution failed:
/home/ubuntu/ros2_ws/src/raspimouse2/raspimouse/src/raspimouse_component.cpp:134:48: note:   ‘std::__cxx11::basic_string<char>’ is not derived from ‘std::chrono::duration<_Rep1, _Period1>’
  134 |       &Raspimouse::publish_light_sensors, this));
      |                                                ^
make[2]: *** [CMakeFiles/raspimouse_component.dir/build.make:63: CMakeFiles/raspimouse_component.dir/src/raspimouse_component.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:107: CMakeFiles/raspimouse_component.dir/all] Error 2
make: *** [Makefile:141: all] Error 2
---
```

以下のように宣言したらビルドは通りました。  
こちらのコミットです 901baed
```cpp
std::chrono::milliseconds ls_hz{static_cast<long int>(get_parameter(LIGHT_SENSORS_HZ_PARAM).get_value<double>())};
```
ただ、ノードをconfigureにしようとすると以下のようなエラーがでます。
```sh
# 実行したコマンド
# Terminal 1
ros2 run raspimouse raspimouse
# Terminal 2
ros2 lifecycle set raspimouse configure
```
```sh
# Terminal 1の実行結果
[INFO] [1638252198.160319747] [raspimouse]: Configuring Raspimouse node
[ERROR] [1638252198.204778679] []: Caught exception in callback for transition 10
[ERROR] [1638252198.204877345] []: Original error: light_sensors_hz
[WARN] [1638252198.204972938] []: Error occurred while doing error handling.
```

## 解決案
追加したls_hz変数より前にdeclare_parameter関数を呼び出す

--- a/README.md
+++ b/README.md
@@ -241,6 +241,15 @@ Similarly other sensor information can also be viewed by echoing the relevant to
 
   Sets the number of pulses needed for 1 rotation of the used motor.
 
+- `light_sensors_hz`
+
+  Type: `double`
+
+  Default: `100.0`
+
+  Sets the frequency of the publishing rate of the topic /light_sensors.
+  The unit is in Hz.
+
 
 ## License
 

--- a/raspimouse/src/raspimouse_component.cpp
+++ b/raspimouse/src/raspimouse_component.cpp
@@ -126,10 +126,10 @@ CallbackReturn Raspimouse::on_configure(const rclcpp_lifecycle::State &)
   switches_timer_->cancel();
   // Timer for publishing light sensor information
   declare_parameter(LIGHT_SENSORS_HZ_PARAM, 10.0);
-  std::chrono::milliseconds ls_hz{static_cast<long int>(get_parameter(LIGHT_SENSORS_HZ_PARAM).get_value<double>())};
+  std::chrono::milliseconds ls_duration{static_cast<long int>(1000.0/get_parameter(LIGHT_SENSORS_HZ_PARAM).get_value<double>())};
   light_sensors_timer_ = create_wall_timer(
   //   100ms, // Original
-    ls_hz, std::bind(
+    ls_duration, std::bind(
       &Raspimouse::publish_light_sensors, this));
   light_sensors_timer_->cancel();
   // Subscriber for LED commands

--- a/raspimouse/src/raspimouse_component.cpp
+++ b/raspimouse/src/raspimouse_component.cpp
@@ -128,7 +128,7 @@ CallbackReturn Raspimouse::on_configure(const rclcpp_lifecycle::State &)
   declare_parameter(LIGHT_SENSORS_HZ_PARAM, 10.0);
   std::chrono::milliseconds light_sensors_duration{static_cast<int64_t>(1000.0/get_parameter(LIGHT_SENSORS_HZ_PARAM).get_value<double>())};
   light_sensors_timer_ = create_wall_timer(
-    ls_duration, std::bind(
+    light_sensors_duration, std::bind(
       &Raspimouse::publish_light_sensors, this));
   light_sensors_timer_->cancel();
   // Subscriber for LED commands

--- a/raspimouse/src/raspimouse_component.cpp
+++ b/raspimouse/src/raspimouse_component.cpp
@@ -193,7 +193,6 @@ CallbackReturn Raspimouse::on_configure(const rclcpp_lifecycle::State &)
   declare_parameter(WHEEL_TREAD_PARAM, 0.0925);
   declare_parameter(PULSES_PER_REVOLUTION_PARAM, 400.0);
 
-
   // Test if the pulse counters are available
   if (get_parameter(use_pulse_counters_param).get_value<bool>()) {
     RCLCPP_INFO(get_logger(), "Testing counters");

--- a/raspimouse/src/raspimouse_component.cpp
+++ b/raspimouse/src/raspimouse_component.cpp
@@ -125,7 +125,7 @@ CallbackReturn Raspimouse::on_configure(const rclcpp_lifecycle::State &)
       &Raspimouse::publish_switches, this));
   switches_timer_->cancel();
   // Timer for publishing light sensor information
-  declare_parameter(LIGHT_SENSORS_HZ_PARAM, 10.0);
+  declare_parameter(LIGHT_SENSORS_HZ_PARAM, 100.0);
   std::chrono::milliseconds light_sensors_duration{static_cast<int64_t>(1000.0/get_parameter(LIGHT_SENSORS_HZ_PARAM).get_value<double>())};
   light_sensors_timer_ = create_wall_timer(
     light_sensors_duration, std::bind(

--- a/raspimouse/src/raspimouse_component.cpp
+++ b/raspimouse/src/raspimouse_component.cpp
@@ -125,7 +125,7 @@ CallbackReturn Raspimouse::on_configure(const rclcpp_lifecycle::State &)
       &Raspimouse::publish_switches, this));
   switches_timer_->cancel();
   // Timer for publishing light sensor information
-  declare_parameter(LIGHT_SENSORS_HZ_PARAM, 100.0);
+  declare_parameter(LIGHT_SENSORS_HZ_PARAM, 10.0);
   std::chrono::milliseconds ls_hz{static_cast<long int>(get_parameter(LIGHT_SENSORS_HZ_PARAM).get_value<double>())};
   light_sensors_timer_ = create_wall_timer(
   //   100ms, // Original

--- a/raspimouse/src/raspimouse_component.cpp
+++ b/raspimouse/src/raspimouse_component.cpp
@@ -33,6 +33,8 @@ constexpr auto use_light_sensors_param = "use_light_sensors";
 constexpr auto odometry_scale_left_wheel_param = "odometry_scale_left_wheel";
 constexpr auto odometry_scale_right_wheel_param = "odometry_scale_right_wheel";
 
+constexpr auto LIGHT_SENSORS_HZ_PARAM = "light_sensors_hz";
+
 constexpr auto WHEEL_DIAMETER_PARAM = "wheel_diameter";
 constexpr auto WHEEL_TREAD_PARAM = "wheel_tread";
 constexpr auto PULSES_PER_REVOLUTION_PARAM= "pulses_per_revolution";
@@ -123,8 +125,10 @@ CallbackReturn Raspimouse::on_configure(const rclcpp_lifecycle::State &)
       &Raspimouse::publish_switches, this));
   switches_timer_->cancel();
   // Timer for publishing light sensor information
+  std::chrono::milliseconds ls_hz{static_cast<long int>(get_parameter(LIGHT_SENSORS_HZ_PARAM).get_value<double>())};
   light_sensors_timer_ = create_wall_timer(
-    100ms, std::bind(
+  //   100ms, // Original
+    ls_hz, std::bind(
       &Raspimouse::publish_light_sensors, this));
   light_sensors_timer_->cancel();
   // Subscriber for LED commands
@@ -188,6 +192,7 @@ CallbackReturn Raspimouse::on_configure(const rclcpp_lifecycle::State &)
   declare_parameter(WHEEL_DIAMETER_PARAM, 0.048);
   declare_parameter(WHEEL_TREAD_PARAM, 0.0925);
   declare_parameter(PULSES_PER_REVOLUTION_PARAM, 400.0);
+  declare_parameter(LIGHT_SENSORS_HZ_PARAM, 100);
 
   // Test if the pulse counters are available
   if (get_parameter(use_pulse_counters_param).get_value<bool>()) {

--- a/raspimouse/src/raspimouse_component.cpp
+++ b/raspimouse/src/raspimouse_component.cpp
@@ -125,6 +125,7 @@ CallbackReturn Raspimouse::on_configure(const rclcpp_lifecycle::State &)
       &Raspimouse::publish_switches, this));
   switches_timer_->cancel();
   // Timer for publishing light sensor information
+  declare_parameter(LIGHT_SENSORS_HZ_PARAM, 100.0);
   std::chrono::milliseconds ls_hz{static_cast<long int>(get_parameter(LIGHT_SENSORS_HZ_PARAM).get_value<double>())};
   light_sensors_timer_ = create_wall_timer(
   //   100ms, // Original
@@ -192,7 +193,7 @@ CallbackReturn Raspimouse::on_configure(const rclcpp_lifecycle::State &)
   declare_parameter(WHEEL_DIAMETER_PARAM, 0.048);
   declare_parameter(WHEEL_TREAD_PARAM, 0.0925);
   declare_parameter(PULSES_PER_REVOLUTION_PARAM, 400.0);
-  declare_parameter(LIGHT_SENSORS_HZ_PARAM, 100);
+
 
   // Test if the pulse counters are available
   if (get_parameter(use_pulse_counters_param).get_value<bool>()) {

--- a/raspimouse/src/raspimouse_component.cpp
+++ b/raspimouse/src/raspimouse_component.cpp
@@ -126,7 +126,7 @@ CallbackReturn Raspimouse::on_configure(const rclcpp_lifecycle::State &)
   switches_timer_->cancel();
   // Timer for publishing light sensor information
   declare_parameter(LIGHT_SENSORS_HZ_PARAM, 10.0);
-  std::chrono::milliseconds ls_duration{static_cast<long int>(1000.0/get_parameter(LIGHT_SENSORS_HZ_PARAM).get_value<double>())};
+  std::chrono::milliseconds light_sensors_duration{static_cast<int64_t>(1000.0/get_parameter(LIGHT_SENSORS_HZ_PARAM).get_value<double>())};
   light_sensors_timer_ = create_wall_timer(
     ls_duration, std::bind(
       &Raspimouse::publish_light_sensors, this));

--- a/raspimouse/src/raspimouse_component.cpp
+++ b/raspimouse/src/raspimouse_component.cpp
@@ -128,7 +128,6 @@ CallbackReturn Raspimouse::on_configure(const rclcpp_lifecycle::State &)
   declare_parameter(LIGHT_SENSORS_HZ_PARAM, 10.0);
   std::chrono::milliseconds ls_duration{static_cast<long int>(1000.0/get_parameter(LIGHT_SENSORS_HZ_PARAM).get_value<double>())};
   light_sensors_timer_ = create_wall_timer(
-  //   100ms, // Original
     ls_duration, std::bind(
       &Raspimouse::publish_light_sensors, this));
   light_sensors_timer_->cancel();

